README.md

```diff
@@ -94,6 +94,32 @@ BBEdit version 14.0 and higher adds support for Language Server Protocols, inclu
 }
 ```
 
+### Emacs `lsp-mode`
+
+Several LSP client implementations for
+[Emacs](https://www.gnu.org/software/emacs/) exist in the form of
+Emacs packages. This installation instruction covers the [LSP
+Mode](https://emacs-lsp.github.io/lsp-mode/) package.
+
+1. Install the the [LSP Mode](https://emacs-lsp.github.io/lsp-mode/)
+   package for Emacs as described in the [`lsp-mode` install
+   instructions](https://emacs-lsp.github.io/lsp-mode/page/installation/).
+
+2. Add the following to your Emacs init file:
+
+   ```
+   (add-to-list 'lsp-language-id-configuration '(perl-mode . "perl-pls"))
```

Comment:
How do you configure all the PLS settings?

Reply:
Depends on what you're looking for. There's command line args (see previous comment), environment vars, and LSP init params.

Full details in the LSP documentation: https://emacs-lsp.github.io/lsp-mode/page/adding-new-language/

Reply:
From looking at the docs for the other Perl language servers, there is some sort of syntax for setting the configuration:

https://emacs-lsp.github.io/lsp-mode/page/lsp-perlnavigator/

I'm guessing a setting like "perl.syntax.enabled" in VSCode would translate to "lsp-perl-syntax-enabled"? Can you test that and document it if that's correct?

Reply:
I looked through the sources of `lsp-mode`, and it seems that for the non-trivial cases, an adapter module for `lsp-mode` needs to be provided. This is for example what the adapter code for PerlNavigator does:
```elisp
(lsp-register-custom-settings
 '(("perlnavigator.trace.server" lsp-perlnavigator-trace-server)
   ("perlnavigator.logging" lsp-perlnavigator-logging t)
   ("perlnavigator.includePaths" lsp-perlnavigator-include-paths)
   ("perlnavigator.severity1" lsp-perlnavigator-severity1)
   ("perlnavigator.severity2" lsp-perlnavigator-severity2)
   ("perlnavigator.severity3" lsp-perlnavigator-severity3)
   ("perlnavigator.severity4" lsp-perlnavigator-severity4)
   ("perlnavigator.severity5" lsp-perlnavigator-severity5)
   ("perlnavigator.perlcriticEnabled" lsp-perlnavigator-perlcritic-enabled t)
   ("perlnavigator.perlcriticProfile" lsp-perlnavigator-perlcritic-profile)
   ("perlnavigator.perltidyProfile" lsp-perlnavigator-perltidy-profile)
   ("perlnavigator.enableWarnings" lsp-perlnavigator-enable-warnings t)
   ("perlnavigator.perlPath" lsp-perlnavigator-perl-path)))
```
Judging from the code that deals with those settings, it seems that the double-quoted string (first element) is parsed, split at full stops (`.`) to create the containment hierarchy, and the value of the elisp variable or function (second element) is evaluated to obtain the value. The result then goes into the initialization options.

My take is that it would seem worth the effort of making such an adapter module, **and to contribute it as a PR to `lsp-mode`.** If we kept the adapter module separate from `lsp-mode`, users would have to go through an extra install step. Given that there are two adapters for Perl in `lsp-mode` already, I would be optimistic that adding a third one will not be controversial.

The other question is, how far you want to go with the present PR, updating the readme? Things do work with the minimal setup given in the readme update, and it seems worthwhile (IMO at least) to contribute a "fuller" version to `lsp-mode` itself.

In this light, how would you prefer to proceed?

Reply:
Yeah, I think it would be best to get a full client implementation into lsp-mode. I'd be happy to do it if you can't, however it might take a bit longer since my emacs is rusty and I'd need to set it up.

Reply:
I just spotted this in the adapter for `Perl::LnaguageServer`:
```elisp
(lsp-register-custom-settings
 '(("perl.perlCmd" lsp-perl-perl-cmd)
   ("perl.perlInc" lsp-perl-perl-inc)
   ("perl.fileFilter" lsp-perl-file-filter)
   ("perl.ignoreDirs" lsp-perl-ignore-dirs)
   ("perl.showLocalVars" lsp-perl-show-local-vars t)
   ("perl.logLevel" lsp-perl-log-level)))
```
This means that those settings could collide (as far as `lsp-mode` is concerned) with your `perl.*` hierarchy.

The reason is that all `lsp-register-custom-settings` stuff is collected in one big list, and then at some point it calls `(lsp-configuration-section "perl")` to collect all settings starting with "perl" for putting them in the init message.

Consequently, the `Perlnavigator` adapter uses its own prefix `perlnavigator` as shown in my previous comment.

Hence, to be able to make an `lsp-mode` adapter for PLS, it would be most helpful to rename the prefix used by PLS.
| Current | Proposed |
|-|-|
| `perl.pls` | `perl.pls.cmd` |
| `perl.plsargs` | `perl.pls.args` |
| `perl.cwd` | `perl.pls.cwd` |
| `perl.inc` | `perl.pls.inc` |
| `perl.perltidyrc` | `perl.pls.perltidyrc` |
| `perl.perlcritic.perlcriticrc` | `perl.pls.perlcritic.perlcriticrc` |
| `perl.perlcritic.enabled` | `perl.pls.perlcritic.enabled` |
| `perl.syntax.enabled` | `perl.pls.syntax.enabled` |

Looking forward to your toughts.

Reply:
> Yeah, I think it would be best to get a full client implementation into lsp-mode. I'd be happy to do it if you can't, however it might take a bit longer since my emacs is rusty and I'd need to set it up.

I'm an Emacs contributor, so will be happy to do the LSP Mode integration.

Reply:
Thanks so much!

Reply:
> [...]
> Realisticaly, no one should ever be using two of these language servers. Even if someone had Perl::LanguageServer settings, it wouldn't really impact PLS since there are no settings with those names. They would all be sent to the server, but they would just be ignored.

True for as long as the folks behind Perl::LanguageServer don't introduce a new parameter that collides with one of yours.

Couldn't you support both, the current, **and** the `perl.pls.` names? Existing configurations would continue to work, and with an appropriate note in the documentation, any new parameters in the future would be added under `perl.pls.` only. Juts my two cents...

Reply:
> Couldn't you support both, the current, and the perl.pls. names? Existing configurations would continue to work, and with an appropriate note in the documentation, any new parameters in the future would be added under perl.pls. only. Just my two cents...

I suppose that's possible. I would need to make sure I *could* have the VSCode extension do a settings migration. Otherwise, it might be a little weird for VSCode users.

Reply:
It looks like the standard thing to do for VSCode is to not remove a setting, but mark it with a deprecation message and point users to the new setting. So I will do that.

Reply:
Deprecating in VSCode would just indicate to users that they should move to the new namespace, and would not show the old settings for new config. The server code would continue to recognize both old and new.

Reply:
Thanks, this clarifies; we're on the same page then.